# Proyecto Final Bienes Raices Mexico 2013-2016

## Librerias

In [1]:
import pandas as pd
import numpy as np

## Importar Datos de Ventas y Rentas de Bienes Raices en Mexico de 2013 a 2016

In [2]:
# Se utiliza el 'latin' encoding ya que contiene descripciones de los inmuebles
data = pd.read_csv('real_estate_mex_2013_2016.csv', encoding = 'latin')
data.head()

created_on operation property_type           place_name  \
0  19/09/2013      sell     apartment  Nuevo Centro Urbano   
1  19/09/2013      sell     apartment  Nuevo Centro Urbano   
2  19/09/2013      sell     apartment  Nuevo Centro Urbano   
3  19/09/2013      sell     apartment     Playa del Carmen   
4  19/09/2013      sell     apartment  Nuevo Centro Urbano   

                             place_with_parent_names  geonames_id  \
0  |México|Quintana Roo|Solidaridad|Playa del Ca...          NaN   
1  |México|Quintana Roo|Solidaridad|Playa del Ca...          NaN   
2  |México|Quintana Roo|Solidaridad|Playa del Ca...          NaN   
3  |México|Quintana Roo|Solidaridad|Playa del Ca...          NaN   
4  |México|Quintana Roo|Solidaridad|Playa del Ca...          NaN   

                lat-lon        lat        lon     price  ...  \
0  20.627527,-87.076686  20.627527 -87.076686  269000.0  ...   
1  20.627567,-87.076718  20.627567 -87.076718  230000.0  ...   
2    20.6267,-87.075866  20.626700 -87.075866  135000.0  ...   
3  20.635482,-87.070345  20.635482 -87.070345  185000.0  ...   
4  20.629092,-87.074658  20.629092 -87.074658  180000.0  ...   

  surface_covered_in_m2  price_usd_per_m2  price_per_m2  floor  rooms  \
0                 110.0       2445.454545   2445.454545    NaN    NaN   
1                 110.0       2090.909091   2090.909091    NaN    NaN   
2                  41.0       3292.682927   3292.682927    NaN    NaN   
3                 122.0       1516.393443   1516.393443    NaN    NaN   
4                   NaN               NaN           NaN    NaN    NaN   

   expenses                                      properati_url  \
0       NaN  http://nuevo-centro-urbano.properati.com.mx/o6...   
1       NaN  http://nuevo-centro-urbano.properati.com.mx/o8...   
2       NaN  http://nuevo-centro-urbano.properati.com.mx/o9...   
3       NaN  http://playa-del-carmen-playa-del-carmen.prope...   
4       NaN  http://nuevo-centro-urbano.properati.com.mx/ok...   

                                         description  \
0  Referencia #28010. En el centro de Playa del C...   
1  Referencia #28925. En el centro de Playa del C...   
2  Referencia #29031. La Papaya un proyecto altam...   
3  Referencia #29032. Visite las nuevas Acqua Ter...   
4  Referencia #33970. Todavía tenemos algunas un...   

                                               title  \
0  Departamento en Venta. Ref. #28010. 3 Dormitor...   
1  Departamento en Venta. Ref. #28925. 3 Dormitor...   
2  Departamento en Venta. Ref. #29031. 1 Dormitor...   
3  Departamento en Venta. Ref. #29032. 2 Dormitor...   
4  Departamento en Venta. Ref. #33970. Encanto Ri...   

                                     image_thumbnail  
0  https://thumbs-cf.properati.com/3/q4Ws9xQT8Yu3...  
1  https://thumbs-cf.properati.com/8/7W4OI5hIXPdY...  
2  https://thumbs-cf.properati.com/0/4h1_ZbVNWly0...  
3  https://thumbs-cf.properati.com/0/Mxv1JulyoQBM...  
4  https://thumbs-cf.properati.com/9/jzkZFLJUFVEw...  

[5 rows x 24 columns]

## Analisis de Tabla de Datos

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188525 entries, 0 to 188524
Data columns (total 24 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   created_on                  188525 non-null  object 
 1   operation                   188525 non-null  object 
 2   property_type               188525 non-null  object 
 3   place_name                  188525 non-null  object 
 4   place_with_parent_names     188525 non-null  object 
 5   geonames_id                 1 non-null       float64
 6   lat-lon                     156591 non-null  object 
 7   lat                         156591 non-null  float64
 8   lon                         156591 non-null  float64
 9   price                       185911 non-null  float64
 10  currency                    185911 non-null  object 
 11  price_aprox_local_currency  185911 non-null  float64
 12  price_aprox_usd             185911 non-null  float64
 13  surface_total_

_Observamos que existen 188,525 registros en total, con variables como fechas, tipo de operaciones, tipo de propiedad, nombre del lugar, descripción más detallada del lugar, variables geograficas, precio, moneda, superficie, piso, numero de habitaciones, descripciones, etc._

## Datos Nulos por Variable

In [4]:
# Suma de Nulos por variable
nulos = pd.DataFrame(data.isnull().sum(), columns = ['Nulos'])
# Agrega porcentaje de nulos con respecto al total de registros
nulos['%_Nulos'] = round(nulos['Nulos']/data.shape[0] * 100, 2).values
nulos

Nulos  %_Nulos
created_on                       0     0.00
operation                        0     0.00
property_type                    0     0.00
place_name                       0     0.00
place_with_parent_names          0     0.00
geonames_id                 188524   100.00
lat-lon                      31934    16.94
lat                          31934    16.94
lon                          31934    16.94
price                         2614     1.39
currency                      2614     1.39
price_aprox_local_currency    2614     1.39
price_aprox_usd               2614     1.39
surface_total_in_m2          90894    48.21
surface_covered_in_m2         7443     3.95
price_usd_per_m2            110834    58.79
price_per_m2                 21170    11.23
floor                       159810    84.77
rooms                       182449    96.78
expenses                    188318    99.89
properati_url                    0     0.00
description                      0     0.00
title                            0     0.00
image_thumbnail               5522     2.93

_Algunas variables como geonames_id que tiene un solo dato o expense con 207 datos seran eliminados ya que contienen mas del 90% de información nula_

## Eliminar Variables con un Porcentaje Mayor del 50% de Informacion Nula

In [5]:
# Nombres de columnas mayores con más del 50% de nulos
col_null = nulos[nulos['%_Nulos'] > 50].index

# Eliminar dichas columnas del data frame original
df = data.drop(col_null, axis = 1)

# Porcentaje de Nulos
round(df.isnull().sum()/df.shape[0] * 100, 2)

created_on                     0.00
operation                      0.00
property_type                  0.00
place_name                     0.00
place_with_parent_names        0.00
lat-lon                       16.94
lat                           16.94
lon                           16.94
price                          1.39
currency                       1.39
price_aprox_local_currency     1.39
price_aprox_usd                1.39
surface_total_in_m2           48.21
surface_covered_in_m2          3.95
price_per_m2                  11.23
properati_url                  0.00
description                    0.00
title                          0.00
image_thumbnail                2.93
dtype: float64

## Analisis y Separación de la Variable place_with_parent_names

In [6]:
df['place_with_parent_names'][0]

'|México|Quintana Roo|Solidaridad|Playa del Carmen|Nuevo Centro Urbano|'

_En la mayoria de los casos los nombres se dividen en país, estados, municipio y colonia_

In [7]:
# Número de Registros en la mayoría de los casos
df['place_with_parent_names'].map(lambda x: len(x.split('|')[1:])).value_counts()

4    139874
5     35146
6      8989
3      4516
Name: place_with_parent_names, dtype: int64

_Se utilizaran los 4 primeros datos en cada uno de los registros, al ser un a analisis de Mexico se eliminaran las columnas con distintos pais, la columna pais sera eliminada_

In [8]:
pais = df['place_with_parent_names'].map(lambda x: x.split('|')[1])
estado = df['place_with_parent_names'].map(lambda x: x.split('|')[2])
municipio = df['place_with_parent_names'].map(lambda x: x.split('|')[3])
colonia = df['place_with_parent_names'].map(lambda x: x.split('|')[4:])

In [9]:
# Relleno de Nan
for i in range(len(colonia)):
    try:
        # Si la colonia es vacia
        if colonia[i][0] == '':
            colonia[i][0] = np.nan
    except:
        # Si no existe la colina
        colonia[i].append(np.nan)

In [10]:
# Solo un elemento
colonia = colonia.map(lambda x: x[0])

In [11]:
# Agregar nuevos elementos al data frame
df['pais'] = pais
df['estado'] = estado
df['municipio'] = municipio
df['colonia'] = colonia

In [12]:
df.head()

created_on operation property_type           place_name  \
0  19/09/2013      sell     apartment  Nuevo Centro Urbano   
1  19/09/2013      sell     apartment  Nuevo Centro Urbano   
2  19/09/2013      sell     apartment  Nuevo Centro Urbano   
3  19/09/2013      sell     apartment     Playa del Carmen   
4  19/09/2013      sell     apartment  Nuevo Centro Urbano   

                             place_with_parent_names               lat-lon  \
0  |México|Quintana Roo|Solidaridad|Playa del Ca...  20.627527,-87.076686   
1  |México|Quintana Roo|Solidaridad|Playa del Ca...  20.627567,-87.076718   
2  |México|Quintana Roo|Solidaridad|Playa del Ca...    20.6267,-87.075866   
3  |México|Quintana Roo|Solidaridad|Playa del Ca...  20.635482,-87.070345   
4  |México|Quintana Roo|Solidaridad|Playa del Ca...  20.629092,-87.074658   

         lat        lon     price currency  ...  surface_covered_in_m2  \
0  20.627527 -87.076686  269000.0      USD  ...                  110.0   
1  20.627567 -87.076718  230000.0      USD  ...                  110.0   
2  20.626700 -87.075866  135000.0      USD  ...                   41.0   
3  20.635482 -87.070345  185000.0      USD  ...                  122.0   
4  20.629092 -87.074658  180000.0      USD  ...                    NaN   

   price_per_m2                                      properati_url  \
0   2445.454545  http://nuevo-centro-urbano.properati.com.mx/o6...   
1   2090.909091  http://nuevo-centro-urbano.properati.com.mx/o8...   
2   3292.682927  http://nuevo-centro-urbano.properati.com.mx/o9...   
3   1516.393443  http://playa-del-carmen-playa-del-carmen.prope...   
4           NaN  http://nuevo-centro-urbano.properati.com.mx/ok...   

                                         description  \
0  Referencia #28010. En el centro de Playa del C...   
1  Referencia #28925. En el centro de Playa del C...   
2  Referencia #29031. La Papaya un proyecto altam...   
3  Referencia #29032. Visite las nuevas Acqua Ter...   
4  Referencia #33970. Todavía tenemos algunas un...   

                                               title  \
0  Departamento en Venta. Ref. #28010. 3 Dormitor...   
1  Departamento en Venta. Ref. #28925. 3 Dormitor...   
2  Departamento en Venta. Ref. #29031. 1 Dormitor...   
3  Departamento en Venta. Ref. #29032. 2 Dormitor...   
4  Departamento en Venta. Ref. #33970. Encanto Ri...   

                                     image_thumbnail     pais        estado  \
0  https://thumbs-cf.properati.com/3/q4Ws9xQT8Yu3...  México  Quintana Roo   
1  https://thumbs-cf.properati.com/8/7W4OI5hIXPdY...  México  Quintana Roo   
2  https://thumbs-cf.properati.com/0/4h1_ZbVNWly0...  México  Quintana Roo   
3  https://thumbs-cf.properati.com/0/Mxv1JulyoQBM...  México  Quintana Roo   
4  https://thumbs-cf.properati.com/9/jzkZFLJUFVEw...  México  Quintana Roo   

     municipio           colonia  
0  Solidaridad  Playa del Carmen  
1  Solidaridad  Playa del Carmen  
2  Solidaridad  Playa del Carmen  
3  Solidaridad  Playa del Carmen  
4  Solidaridad  Playa del Carmen  

[5 rows x 23 columns]

In [13]:
# Paises distitnos de Mexico
df['pais'].value_counts()

México                       188068
Brasil                           456
Estados Unidos de América         1
Name: pais, dtype: int64

In [14]:
# Eliminar Paises distinto de Mexico
df = df[df['pais'] == 'México'].reset_index(drop = True)

# Eliminar Columna pais ya que todo son Mexico
df.drop(['pais'], axis = 1, inplace = True)
# Eliminar columna place_with_parent_names y place_name
df.drop(['place_with_parent_names', 'place_name'], axis = 1, inplace = True)

## Eliminar Variable lat-lon

_Esta variable sera eliminada ya que existen variables independientes con la misma información_

In [15]:
df.iloc[:, 3:5].isnull().sum()

lat-lon    31934
lat        31934
dtype: int64

_Observamos la misma cantidad de null en ambas variables, lo que indica que estan relacionadas_

In [16]:
# Elimnar lat-lon
df.drop(['lat-lon'], axis = 1, inplace = True)

## Cambio de Nombres de Variables

In [17]:
# Observamos el nombre de las variables
list(df.columns)

['created_on',
 'operation',
 'property_type',
 'lat',
 'lon',
 'price',
 'currency',
 'price_aprox_local_currency',
 'price_aprox_usd',
 'surface_total_in_m2',
 'surface_covered_in_m2',
 'price_per_m2',
 'properati_url',
 'description',
 'title',
 'image_thumbnail',
 'estado',
 'municipio',
 'colonia']

In [18]:
continuas = ['lat', 'lon', 'price', 'price_aprox_local_currency', 'price_aprox_usd',
            'surface_total_in_m2', 'surface_covered_in_m2', 'price_per_m2']
discretas = ['operation', 'property_type', 'currency', 'estado', 'municipio', 'colonia']
fechas = ['created_on']
text = ['properati_url', 'description', 'title', 'image_thumbnail']

In [19]:
# Renombrar variables continuas
df.rename(columns = dict(zip(continuas, list(map(lambda x: 'c_' + x, continuas)))), inplace = True)
# Renombrar variables discretas
df.rename(columns = dict(zip(discretas, list(map(lambda x: 'v_' + x, discretas)))), inplace = True)
# Renombrar variables fecha
df.rename(columns = dict(zip(fechas, list(map(lambda x: 'd_' + x, fechas)))), inplace = True)
# Renombrar variables continuas
df.rename(columns = dict(zip(text, list(map(lambda x: 't_' + x, text)))), inplace = True)